src/connectors/toggle-refinement/connectToggleRefinement.js

```diff
@@ -279,6 +181,124 @@ export default function connectToggleRefinement(renderFn, unmountFn = noop) {
         return state.removeDisjunctiveFacet(attribute);
       },
 
+      getRenderState(renderState, renderOptions) {
+        return {
+          ...renderState,
+          toggleRefinement: this.getWidgetRenderState(renderOptions),
+        };
+      },
+
+      getWidgetRenderState({ state, helper, results, createURL }) {
+        let value;
+
+        // results are null at init
+        if (results) {
+          const isRefined =
+            on &&
+            on.every(v => helper.state.isDisjunctiveFacetRefined(attribute, v));
+          const offValue = toArray(off === undefined ? false : off);
+          const allFacetValues = results.getFacetValues(attribute) || [];
+
+          const onData =
+            on &&
+            on
+              .map(v =>
+                find(
+                  allFacetValues,
+                  ({ name }) => name === unescapeRefinement(v)
+                )
+              )
+              .filter(v => v !== undefined);
+          const onFacetValue = {
+            isRefined:
+              onData.length > 0 ? onData.every(v => v.isRefined) : false,
+            count:
+              onData.length === 0
+                ? null
+                : onData.reduce((acc, v) => acc + v.count, 0),
+          };
+
+          const offData = hasAnOffValue
+            ? offValue
+                .map(v =>
+                  find(
+                    allFacetValues,
+                    ({ name }) => name === unescapeRefinement(v)
+                  )
+                )
+                .filter(v => v !== undefined)
+            : [];
+          const offFacetValue = {
+            isRefined:
+              offData.length > 0 ? offData.every(v => v.isRefined) : false,
+            count:
+              offData.length === 0
+                ? allFacetValues.reduce((total, { count }) => total + count, 0)
+                : offData.reduce((acc, v) => acc + v.count, 0),
+          };
+
+          // what will we show by default,
+          // if checkbox is not checked, show: [ ] free shipping (countWhenChecked)
+          // if checkbox is checked, show: [x] free shipping (countWhenNotChecked)
+          const nextRefinement = isRefined ? offFacetValue : onFacetValue;
+
+          value = {
+            name: attribute,
+            isRefined,
+            count: nextRefinement === undefined ? null : nextRefinement.count,
+            onFacetValue,
+            offFacetValue,
+          };
+        } else {
+          const isRefined =
```

Comment:
some of the code is now duplicated in both branches, and could put put top-level and deduplicated, right?

Reply:
I might be able to group `onFacetValue` and `offFacetValue` with a spicy ternary, I'll try that!

Reply:
https://github.com/algolia/instantsearch.js/pull/4569#discussion_r516802357